[RISCV] Allow hoisting VXRM writes out of loops speculatively

felipealmeida opened the pull request:
Change the intersect for the anticipated algorithm to ignore unknown when anticipating. This effectively allows VXRM writes speculatively because it could do a VXRM write even when there's branches where VXRM is unneeded.

The importance of this change is because VXRM writes causes pipeline flushes in some micro-architectures and so it makes sense to allow more aggressive hoisting even if it causes some degradation for the slow path.

An example is this code:
```
typedef unsigned char uint8_t;
__attribute__ ((noipa))
void foo (uint8_t *dst,  int i_dst_stride,
           uint8_t *src1, int i_src1_stride,
           uint8_t *src2, int i_src2_stride,
           int i_width, int i_height )
{
   for( int y = 0; y < i_height; y++ )
     {
       for( int x = 0; x < i_width; x++ )
         dst[x] = ( src1[x] + src2[x] + 1 ) >> 1;
       dst  += i_dst_stride;
       src1 += i_src1_stride;
       src2 += i_src2_stride;
     }
}
```
With this patch, the code above generates a hoisting VXRM writes out of the outer loop.

felipealmeida committed:
[RISCV] Allow hoisting VXRM writes out of loops speculatively

Change the intersect for the anticipated algorithm to ignore unknown
when anticipating. This effectively allows VXRM writes speculatively
because it could do a VXRM write even when there's branches where VXRM
is unneeded.

The importance of this change is because VXRM writes causes pipeline
flushes in some micro-architectures and so it makes sense to allow
more aggressive hoisting even if it causes some degradation for the
slow path.

An example is this code:
```
typedef unsigned char uint8_t;
__attribute__ ((noipa))
void foo (uint8_t *dst,  int i_dst_stride,
           uint8_t *src1, int i_src1_stride,
           uint8_t *src2, int i_src2_stride,
           int i_width, int i_height )
{
   for( int y = 0; y < i_height; y++ )
     {
       for( int x = 0; x < i_width; x++ )
         dst[x] = ( src1[x] + src2[x] + 1 ) >> 1;
       dst  += i_dst_stride;
       src1 += i_src1_stride;
       src2 += i_src2_stride;
     }
}
```
With this patch, the code above generates a hoisting VXRM writes out of the outer loop.

JeffreyALaw commented:
So just to chime in since I suggested Felipe tackle this problem.

As Felipe noted some designs flush their pipelines on a VXRM write which can be a significant performance issue.  I measured a 2-3% improvement with a patch tackling this problem for GCC running spec2017's x264 workloads on the BPI-F3 board.   That all comes from speculative hoisting of VXRM assignments needed to use vaaddu to implement the ceiling average found in pixel_avg.

topperc commented:
> So just to chime in since I suggested Felipe tackle this problem.
> 
> As Felipe noted some designs flush their pipelines on a VXRM write which can be a significant performance issue. I measured a 2-3% improvement with a patch tackling this problem for GCC running spec2017's x264 workloads on the BPI-F3 board. That all comes from speculative hoisting of VXRM assignments needed to use vaaddu to implement the ceiling average found in pixel_avg.

Do you have data for not using vaaddu in that loop and using a vwadd.vv, a vadd.vi, and a vnsrl instead? So that there's no VXRM write needed?

SiFive's p470 and p670 don't optimize VXRM writes well either. Even with the writes hoisted of the loop nest, I couldn't get the vaaddu to be profitable.

JeffreyALaw commented:
> > So just to chime in since I suggested Felipe tackle this problem.
> > As Felipe noted some designs flush their pipelines on a VXRM write which can be a significant performance issue. I measured a 2-3% improvement with a patch tackling this problem for GCC running spec2017's x264 workloads on the BPI-F3 board. That all comes from speculative hoisting of VXRM assignments needed to use vaaddu to implement the ceiling average found in pixel_avg.
> 
> Do you have data for not using vaaddu in that loop and using a vwadd.vv, a vadd.vi, and a vnsrl instead? So that there's no VXRM write needed?
> 
> SiFive's p470 and p670 don't optimize VXRM writes well either. Even with the writes hoisted of the loop nest, I couldn't get the vaaddu to be profitable.

No.    We largely set this aside as we adjusted the Ventana design to make VXRM assignments cheap.   But it seemed a shame to leave things performing so badly on the BPI given we had bits to make it "no so bad".

JeffreyALaw commented:
So pondering this overnight, obviously it's up to y'all what to do for LLVM around this specific patch.

However it does seem clear to me that we're going to have designs with wildly different performance characteristics for VXRM assignments and thus selecting vaaddu and friends really needs to be a micro-arch tuning decision.  I'll have to get the ball rolling on that for GCC.

topperc commented on llvm/lib/Target/RISCV/RISCVInsertWriteVXRM.cpp:
```diff
@@ -109,6 +109,35 @@ class VXRMInfo {
+    // If either is unknown, the result is the other one.
+    if (isUnknown())
+      return Other;
+    else if (Other.isUnknown())
```
No `else` since the `if` always returns. https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return

topperc commented on llvm/test/CodeGen/RISCV/rvv/vxrm-insert-out-of-loop.ll:
```diff
@@ -0,0 +1,386 @@
+
+
+; test1
+define dso_local void @test1(ptr nocapture noundef writeonly %dst, i32 noundef signext %i_dst_stride, ptr nocapture noundef readonly %src1, i32 noundef signext %i_src1_stride, ptr nocapture noundef readonly %src2, i32 noundef signext %i_src2_stride, i32 noundef signext %i_width, i32 noundef signext %i_height) local_unnamed_addr #0 {
```
Drop `dso_local`, `local_unnamed_addr`, and `#0`

felipealmeida commented:
Hello,

I added a TuneVXRMPipelineFlush subfeature and conditioned the changes to VXRM on this. I added this subtarget feature to P470 and P670 and changed the tests accordingly.

I didn't find the SpacemiT K1 processor definition for the BPI. And probably other CPUs could benefit from this as well. Let me know if I should add that to any other processors.

topperc commented:
> Hello,
> 
> I added a TuneVXRMPipelineFlush subfeature and conditioned the changes to VXRM on this. I added this subtarget feature to P470 and P670 and changed the tests accordingly.
> 
> I didn't find the SpacemiT K1 processor definition for the BPI. And probably other CPUs could benefit from this as well. Let me know if I should add that to any other processors.

`def SPACEMIT_X60 : RISCVProcessorModel<"spacemit-x60"` is the BPI

topperc commented:
reverse ping

felipealmeida commented:
Added X60

topperc reviewed (approved):
LGTM

zqb-all commented:
LGTM

felipealmeida commented:
Rebased and fixed test from changes to X60 tuning

topperc commented:
@felipealmeida do you need me to merge this?

felipealmeida commented:
> @felipealmeida do you need me to merge this?

Yes, please. I'm not a committer.
